### PR TITLE
Nfukuzaki/returned value

### PR DESCRIPTION
**Pull request by nfukuzaki**

### CODE REVIEW

File: ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift [hunk 1]
```diff
@@ -35,39 +35,188 @@ public class TriggeredServerCodeResult: NSObject, NSCoding {
...
+        if let str = self.returnedValue as? String {
+            return str
+        }
+        if let num = self.returnedValue as? NSNumber {
+            return String(num)
+        }
+        return nil
+    }
+    public func getReturnedValueAsBool() -> Bool? {
+        if let bool = self.returnedValue as? Bool {
```
Comment (rizakii) [inline ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift]: you don't have to check this

``` swift
return self.returnedValue as? Bool
```

will return `nil` if returned value is not Bool

File: ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift [hunk 2]
```diff
@@ -35,39 +35,188 @@ public class TriggeredServerCodeResult: NSObject, NSCoding {
...
         self.returnedValue = returnedValue
         self.executedAt = executedAt
-        self.errorMessage = errorMessage
+        self.error = error
+    }
+    
+    public func getReturnedValue() -> AnyObject? {
+        return self.returnedValue
+    }
+    public func getReturnedValueAsString() -> String? {
```
Comment (yongpingchen) [inline ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift:50]: I think only for string type is clear. What do you think?

File: ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift [hunk 3]
```diff
@@ -35,39 +35,188 @@ public class TriggeredServerCodeResult: NSObject, NSCoding {
...
+        if error != nil {
+            serverError = ServerError.errorWithNSDict(error!)
+        }
 
         let executedAt = NSDate(timeIntervalSince1970: (executedAtStamp?.doubleValue)!/1000.0)
         
         var result: TriggeredServerCodeResult?
         if succeeded != nil {
-            result = TriggeredServerCodeResult(succeeded:succeeded!, returnedValue:returnedValue, executedAt:executedAt, errorMessage:errorMessage)
+            result = TriggeredServerCodeResult(succeeded:succeeded!, returnedValue:returnedValue, executedAt:executedAt, error:serverError)
```
Comment (yongpingchen) [inline ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift]: `succeeded` and `executedAt` are required NonNullable, better to guard them at the beginning of this func.

File: ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift [hunk 4]
```diff
@@ -35,39 +35,188 @@ public class TriggeredServerCodeResult: NSObject, NSCoding {
...
+            } else if e1 is NSNumber {
+                if e1 as! NSNumber != e2 as! NSNumber {
+                    return false
+                }
+            } else {
+                return false
+            }
+        }
+        return true
     }
```
Comment (yongpingchen) [inline ThingIFSDK/ThingIFSDK/TriggeredServerCodeResult.swift:144]: hmm, `isEqual` is complex. Please add tests for it.

**Comment by yongpingchen** [comment]
@nfukuzaki please confirm the comments

**Comment by yongpingchen** [comment]
LGTM

**Pull request merged**